rust/template/distributed_datalog/src/schema/config.rs

```diff
@@ -0,0 +1,140 @@
+    // #[cfg(test)]
+    // Observable(observable: Box<dyn Observable>),
+}
+
+impl Ord for Source {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self {
+            Source::File(file) => match other {
+                Source::File(other_file) => file.cmp(other_file),
+            },
+        }
+    }
+}
+
+impl PartialOrd for Source {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// All the output sinks we support.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub enum Sink {
+    /// Output is emitted into a file.
+    File(PathBuf),
+    // /// An observer to emit output to directly.
+    // ///
+    // /// This variant is only meant for testing purposes.
+    // #[cfg(test)]
+    // Observer(observer: ObserverBox<T, E>),
+}
+
+impl Ord for Sink {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self {
+            Sink::File(file) => match other {
+                Sink::File(other_file) => file.cmp(other_file),
+            },
+        }
+    }
+}
+
+impl PartialOrd for Sink {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A description of inputs and outputs of a relation.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub enum RelInOut {
+    /// A relation that is directly fed from a source.
+    Source(Source),
+    /// This relation is fed from the given set of relations.
+    Input(Relations),
+    /// A relation that outputs directly into a sink.
+    Sink(Sink),
+}
+
+impl Ord for RelInOut {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self {
+            RelInOut::Source(source) => match other {
+                RelInOut::Source(other_source) => source.cmp(other_source),
+                RelInOut::Input(_) => Ordering::Less,
+                RelInOut::Sink(_) => Ordering::Less,
+            },
+            RelInOut::Input(rels) => match other {
+                RelInOut::Source(_) => Ordering::Greater,
+                RelInOut::Input(other_rels) => rels.cmp(other_rels),
+                RelInOut::Sink(_) => Ordering::Less,
+            },
+            RelInOut::Sink(sink) => match other {
+                RelInOut::Source(_) => Ordering::Greater,
+                RelInOut::Input(_) => Ordering::Greater,
+                RelInOut::Sink(other_sink) => sink.cmp(other_sink),
+            },
+        }
+    }
+}
+
+impl PartialOrd for RelInOut {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Configuration of input/output relationships for a single node.
+pub type InOut = BTreeMap<RelId, BTreeSet<RelInOut>>;
+
+/// The configuration of the system is comprised of input/output
+/// configurations for a certain set of nodes.
+///
+/// Each element in the set represents a single node.
+pub type Config = BTreeSet<InOut>;
```

Comment:
Shouldn't this be a map from node id's?

Reply:
Potentially. Right now there is nothing that is specific about a node. All are made equal. As such, I wouldn't know what to use as key.
In my current thinking that can't be the node's address, because the assignment of a part of the computation to a member is done as part of some computation in this file (see `assign` function).